### PR TITLE
first try

### PR DESCRIPTION
**Pull request by YevheniiLohatskyi**

### CODE REVIEW

File: src/com/homework/MyArrayList.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    public void add(E e) {
+        if (lastElementIndex >= CAPACITY-1) {
+            CAPACITY = (CAPACITY * 3) / 2 + 1;
+        }
+        array[++lastElementIndex] = e;
+    }
+
+    @Override
+    public void remove(int index) {
+        for (int i = index; i < lastElementIndex; i++) {
```
Comment (Anderwerty) [inline src/com/homework/MyArrayList.java]: The complexity of the algorithm O (n). Try to make the complexity O(1)

File: src/com/homework/MyArrayList.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Override
+    public int size() {
+        return lastElementIndex + 1;
+    }
+
+    @Override
+    public void clear() {
+        array = new MyArrayList<E>().array;
```
Comment (Anderwerty) [inline src/com/homework/MyArrayList.java]: Why not just this.array =new Object[CAPACITY]; ?

File: src/com/homework/MyLinkedList.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public void clear() {
+        root = null;
+    }
+
+    public class Node {
+        private E value;
+        Node next, prev;
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: It is not a java style. 
Every variable declaration (field or local) declares only one variable: declarations such as int a, b; are not used.
https://mate-academy.github.io/style-guides/java/java.html#s5-naming

Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: Why not private?

File: src/com/homework/MyLinkedList.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        node = root;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+    }
+
+    @Override
+    public int size() {
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: It very difficult algorithm. Use field like it ArrayList realization for size?

File: src/com/homework/MyLinkedList.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
+package com.homework;
+
+public class MyLinkedList<E> implements List<E> {
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: Try to use this class and find out it all functional work good or not?

File: src/com/homework/MyLinkedList.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    private Node root;
+
+    @Override
+    public E get(int index) {
+        Node node;
+        if (index < 0) return null;
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: Braces should be added.

File: src/com/homework/MyLinkedList.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        node = root;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node.getValue();
+    }
+
+    @Override
+    public void add(E e) {
+        Node node = new Node(e), tmp;
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: https://mate-academy.github.io/style-guides/java/java.html#s4.8.2-variable-declarations

File: src/com/homework/MyLinkedList.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
+package com.homework;
+
+public class MyLinkedList<E> implements List<E> {
+    private Node root;
+
+    @Override
+    public E get(int index) {
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: What if index>size-1?

File: src/com/homework/MyLinkedList.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+            tmp = tmp.next;
+        }
+        tmp.next = node;
+        node.prev = tmp;
+    }
+
+    @Override
+    public void remove(int index) {
+        Node node;
+        if (index < 0) return;
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: Braces?

File: src/com/homework/MyStack.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
+package com.homework;
+
+public class MyStack<E>{
```
Comment (Anderwerty) [inline src/com/homework/MyStack.java]: `Try to use this class. Fin out all bugs and fix them?

File: src/com/homework/MyArrayList.java (hunk 1 above)
Comment (YevheniiLohatskyi) [reply src/com/homework/MyArrayList.java]: But it's impossible in ArrayList, O(1) can be achieved only in LinkedList

Comment (Anderwerty) [reply src/com/homework/MyArrayList.java]: Really? What about System.arraycopy().

Comment (YevheniiLohatskyi) [reply src/com/homework/MyArrayList.java]: Oh, I understand now, thanks

File: src/com/homework/MyLinkedList.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+public class MyLinkedList<E> implements List<E> {
+    private Node root;
+    private int countElements = 0;
+
+    @Override
+    public E get(int index) {
+        Node node;
+        if (index < 0 || index >= countElements) {
+            return null;
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: Null could be the value of the Node. Better to throw a custom exception.

File: src/com/homework/MyLinkedList.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
+package com.homework;
+
+public class MyLinkedList<E> implements List<E> {
+    private Node root;
+    private int countElements = 0;
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: size better name

File: src/com/homework/MyLinkedList.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Override
+    public void clear() {
+        countElements = 0;
+        root = null;
+    }
+
+    private class Node {
+        private E value;
+        Node next;
```
Comment (Anderwerty) [inline src/com/homework/MyLinkedList.java]: Just explain; why not private?

Comment (YevheniiLohatskyi) [reply src/com/homework/MyLinkedList.java]: I thought you were talking about the class. Will fix it

File: src/com/homework/MyLinkedList.java (hunk 5 above)
Comment (Anderwerty) [reply src/com/homework/MyLinkedList.java]: Please read https://habr.com/ru/post/127864/ Add fix comments.

File: src/com/homework/MyArrayList.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    private int lastElementIndex = -1;
+    private Object[] array;
+
+    public MyArrayList() {
+        array = new Object[capacity];
+    }
+
+    @Override
+    public E get(int index) {
+        return (index >= 0 && index <= lastElementIndex) ? (E) array[index] : null;
```
Comment (Anderwerty) [inline src/com/homework/MyArrayList.java]: If index >= 0 && index <= lastElementIndex it better to throw an exception not to return null. As an element in the list could be null. And your user won't be able to understand if the element is null or index is not right.

File: src/com/homework/MyLinkedList.java (hunk 11 above)
Comment (Anderwerty) [reply src/com/homework/MyLinkedList.java]: The same.

**Pull request closed without merging**